### ISSUE TITLE
Disabling automatic .train() for loss criteria 

### ISSUE THREAD

**Issue by eladrich** [PyTorchLightning/pytorch-lightning#2824]
#### What is your question?
I have a loss module that is loaded as part of my lightning module with its own inner network. (output is passed through the network and the result is used to compute the loss) 

The problem is that when starting a train_step Lightning automatically changes the entire module to .train() which recursively changes the loss inner network as well. As the inner network has batch normalization layers this effectively makes the loss criteria change during training which is of course not desired. 

Is there a right way to incorporate such a loss function?

**Comment by nateraw** [comment]
Can you give me an example of what this looks like? Perhaps in a colab notebook?

**Comment by eladrich** [comment]
Sure, the full code is a bit involved, but here is a basic example

```python
class FeatureLoss(nn.Module):
    def __init__(self, ckpt_path):
        super(FeatureLoss, self).__init__()
        self.facenet = LossArch()
        self.facenet.load_state_dict(torch.load(ckpt_path))

    def forward(self, y_hat, y):
        y_hat_feats = self.facenet(y_hat)
        y_feats = self.facenet(y)  
        return F.mse_loss(y_heat_feats, y_feats)

class CoreNet(pl.LightningModule):

    def __init__(self, hparams):
        super(CoreNet, self).__init__()

        # Init Network
        self.net = MyArch()

        # Load loss
        self.feat_loss = FeatureLoss('path_to_weights.ckpt')

    def forward(x):
        return self.net(x)

    def training_step(self, batch):
        x, y = batch
        y_hat = self(x) #.net is in .train() mode
        loss = self.feat_loss(y, y_hat)        # loss is also in .train() mode 
      
```

**Comment by SkafteNicki** [comment]
Two options on top of my mind:
1. You need to make sure that your loss module is not registered as a child module of your network. If the loss module is not registered as a child module then `.train()` will have no effect. Easiest way to do this is to wrap in a list: `self.feat_loss = [FeatureLoss('path_to_weights.ckpt')]`
2. Overwrite `.train` method of `FeatureLoss` so it does nothing. When `CoreNet.train()` is called it recursively call the `.train()` method of all its child modules, so if `FeatureLoss.train()` does nothing then it would not change.

**Comment by eladrich** [comment]
Nice ideas, what I'm currently doing is to manually change `self.feat_loss` to `.eval()` in each training_step, so these would probably be more elegant. 

I think that `1.` is probably a better way to go because it also stops lightning from saving the weights of the loss in each checkpoint which also bothered me.

However, it still feels a bit hacky and I think that especially  for new practitioners it would be hard to figure out that `lightning` might change their loss modules to `.train()` mode, as the `LightningModule __init__` still seems like the place the loss initialization needs to go in `lightning` code structure.

**Comment by SkafteNicki** [comment]
I agree that it may feel a bit hacky, but I would say that it has nothing to do with lightning framework and more to do with native pytorch. You would have the same problem whether or not your `CoreNet` is a `LightningModule` or a `nn.Module` (that calling `.train` on the outermost module will alter all submodules)

**Comment by eladrich** [comment]
I don't totally agree. In native `PyTorch` I will never put the loss module as part of my network. It would be a different object that exists in the main training loop context and therefore I would be able to pass the network from `.train` to `.eval` without affecting it.

Unless I'm missing something in `Lighthning`'s structure, I don't have a different location to place the loss module as it has to be accessed in the `training_step`, and this is what causes the problems here. So the problem here, in my opinion, is caused by the fact that the trainable network is the same object that defines the training loop and loss.

**Comment by SkafteNicki** [comment]
Fair point. You are not missing something in `Lightning`s struture, and I think your problem is very application specific. I am not familiar with this approach where you want to keep some of the trainable network in `eval` mode during training. 

This maybe hints that you could just subclass your normalization layers such that they do not change when `.train()` is called.

**Comment by eladrich** [comment]
I think it happens whenever a loss function is based on a trainable network, for example when using facial identity loss or perceptual loss.

**Comment by SkafteNicki** [comment]
Alright, I better understand the setting now. I think the compromise for now is to see if more people are interested in such a feature and then take it from there. 

Our documentation can always be made clearer or extended (it is implicitly shown in the figure on this page https://pytorch-lightning.readthedocs.io/en/latest/trainer.html that `model.train()` is called inside trainer) so if you have a proposed way to improve this feel free to submit a PR.

**Comment by orpatashnik** [comment]
Agree with @eladrich that this is an important feature. 
I use perceptual loss which depends on a network.

**Comment by zhen8838** [comment]
a simple solution

```python
class FeaturesNet(pl.LightningModule):
  def __init__(self, weights_path):
    super().__init__()
    self.model = MobileFaceNet(512)
    self.model.load_state_dict(torch.load(weights_path))

  def setup(self, device: torch.device):
    self.freeze()

  def train(self, mode: bool):
    """ avoid pytorch lighting auto set trian mode """
    return super().train(False)

  def state_dict(self, destination, prefix, keep_vars):
    """ avoid pytorch lighting auto save params """
    destination = OrderedDict()
    destination._metadata = OrderedDict()
    return destination

```

**Comment by zhen8838** [comment]
@SkafteNicki I think no need to add this feature in PyTorch lighting but need add more detail in [transfer learning document](https://pytorch-lightning.readthedocs.io/en/latest/transfer_learning.html#example-imagenet-computer-vision). When I training a GAN model, use the pre-trained module for content loss, but the document didn't say the model will automatically switch to training mode.

**Comment by altmann144** [comment]
> a simple solution
> 
> ```python
> class FeaturesNet(pl.LightningModule):
>   def __init__(self, weights_path):
>     super().__init__()
>     self.model = MobileFaceNet(512)
>     self.model.load_state_dict(torch.load(weights_path))
> 
>   def setup(self, device: torch.device):
>     self.freeze()
> 
>   def train(self, mode: bool):
>     """ avoid pytorch lighting auto set trian mode """
>     return super().train(False)
> 
>   def state_dict(self, destination, prefix, keep_vars):
>     """ avoid pytorch lighting auto save params """
>     destination = OrderedDict()
>     destination._metadata = OrderedDict()
>     return destination
> ```

thanks, helped a lot
Does not change the functionality, but setup() takes type str for stage, nut torch.device
And as I did the error: use setup in the CoreNet analogue in respect to the first example

**Comment by zhen8838** [comment]
@altmann144 you call `setup` int `on_fit_start`:
```python
class MainModel(pl.LightningModule):
  def __init__(
      self,  **kwargs):
    super().__init__()
    self.save_hyperparameters()

    # networks
    self.generator = xxx
    self.pretrained = self.FeaturesNet()

  def on_fit_start(self) -> None:
    self.pretrained.setup(self.device)

  def forward(self, im):
    return self.generator(im)
```

**Comment by rubencart** [comment]
@SkafteNicki I spent a lot of time looking for a reason for decreased performance, turns out it was because PL put my pretrained (frozen) RNN/CNN into train mode (after I put them in eval mode). I agree that it makes sense to put the model to be trained into train mode. But maybe documentation could me made clearer on the fact that this also happens with frozen models (that you use to encode your input or in a loss or something)? I know that this is only logical, but it is easy to overlook, and maybe a warning somewhere would have helped :). This seems a pretty common use case to me. 

Or, maybe there could be a way to tell PL which models you want to train and which not? So only the ones to be trained would be put in train mode? Or, the `.train()` could maybe be skipped for modules that have zero parameters that require a gradient (so that are frozen)?

**Comment by etienne87** [comment]
Also for Perceptual Loss, how to avoid the module to be automatically saved ?

**Comment by altmann144** [comment]
> a simple solution
> 
> ```python
> class FeaturesNet(pl.LightningModule):
>   def __init__(self, weights_path):
>     super().__init__()
>     self.model = MobileFaceNet(512)
>     self.model.load_state_dict(torch.load(weights_path))
> 
>   def setup(self, device: torch.device):
>     self.freeze()
> 
>   def train(self, mode: bool):
>     """ avoid pytorch lighting auto set trian mode """
>     return super().train(False)
> 
>   def state_dict(self, destination, prefix, keep_vars):
>     """ avoid pytorch lighting auto save params """
>     destination = OrderedDict()
>     destination._metadata = OrderedDict()
>     return destination
> ```

@etienne87 
here with the code from zhen8838, just make sure to use the `state_dict` definition from there.
His comment there is the also answer to your question.

**Comment by etienne87** [comment]
ah ok thanks!

**Comment by etienne87** [comment]
Ah wait, not it does not work when you resume:
`        Missing key(s) in state_dict: "vgg_perc_l1.mean", "vgg_perc_l1.std", "vgg_perc_l1.blocks.0.0.
weight", "vgg_perc_l1.blocks.0.0.bias", "vgg_perc_l1.blocks.0.2.weight", "vgg_perc_l1.blocks.0.2.bias
", "vgg_perc_l1.blocks.1.5.weight", "vgg_perc_l1.blocks.1.5.bias", "vgg_perc_l1.blocks.1.7.weight", "
vgg_perc_l1.blocks.1.7.bias", "vgg_perc_l1.blocks.2.10.weight", "vgg_perc_l1.blocks.2.10.bias", "vgg_
perc_l1.blocks.2.12.weight", "vgg_perc_l1.blocks.2.12.bias", "vgg_perc_l1.blocks.2.14.weight", "vgg_p
erc_l1.blocks.2.14.bias", "vgg_perc_l1.blocks.3.17.weight", "vgg_perc_l1.blocks.3.17.bias", "vgg_perc
_l1.blocks.3.19.weight", "vgg_perc_l1.blocks.3.19.bias", "vgg_perc_l1.blocks.3.21.weight", "vgg_perc_
l1.blocks.3.21.bias".
`

i think the correct solution is rather to use 'on_save_checkpoint' and 'on_load_checkpoint' to specify which model you want to save in the state_dict

**Comment by altmann144** [comment]
Yes you also need to load the params for your model separately from the perceptual loss, meaning loading perceptual loss params in the init of your perceptual loss. That way you separate the checkpoint into the model you are training and the perceptual loss. This makes your whole model more modular and the learned model independent of the used perceptual loss (useful for Inference later)

**Comment by kfarivar** [comment]
Hi, I'm trying to use an encoder (Resnet50) that is supposed to stay fixed during the training of the linear layer I add on top. Is this still the best solution ?

>     1. You need to make sure that your loss module is not registered as a child module of your network. If the loss module is not registered as a child module then `.train()` will have no effect. Easiest way to do this is to wrap in a list: `self.feat_loss = [FeatureLoss('path_to_weights.ckpt')]`

**Comment by ahahahahawd** [comment]
Hi, maybe we can fix it through torchmetrics,  do you have any ideas?
